Web Inspector: Multiline REPL editor is broken

rcaliman-apple opened the pull request:
#### 0da0eedeaa3f18bfd0bb2f1f4831f4fe3eaa4893
<pre>
Web Inspector: Multiline REPL editor is broken
<a href="https://bugs.webkit.org/show_bug.cgi?id=276530">https://bugs.webkit.org/show_bug.cgi?id=276530</a>
<a href="https://rdar.apple.com/131756916">rdar://131756916</a>

Reviewed by BJ Burg and Devin Rousso.

The content area of the console prompt is controlled by CodeMirror
which sets the dimensions of a multi-line input as you type.
It lives in a second level flex container and is intended to be capped at 33vh.

At the first level flex container, the space available to be distributed to flex items
is influenced by the size of the Console Drawer and the min-width of #tab-browser.

Due to alignment within the flex layout and insufficient space,
the multi-line console prompt nested within the second level flex container
was allowed to grow within the second-level container without causing overflow.
This masks some keyboard input and focus.

This patch ensures the maximum height of the Console Drawer takes into account the
space used by the console prompt. It also moves the 33vh max-height limit
to the first level flex container to enforce the limit and ensure a scroll container
as soon as its contents overflow.

* Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js:
(WI.ConsoleDrawer.prototype._updateDrawerHeight):
* Source/WebInspectorUI/UserInterface/Views/Main.css:
(#content):
* Source/WebInspectorUI/UserInterface/Views/QuickConsole.css:
(.quick-console):
(.quick-console &gt; .console-prompt):

Canonical link: <a href="https://commits.webkit.org/283412@main">https://commits.webkit.org/283412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db1649145dcc6b47c8681cdcc50247f42928c612

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45595 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70254 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16832 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53394 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17113 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11729 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69289 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42053 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/57335 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33772 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38724 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14715 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15708 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15060 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71957 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10177 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14452 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60454 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60754 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14593 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8404 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41403 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42479 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43662 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42223 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/QuickConsole.css:
```diff
@@ -51,7 +52,8 @@
 
-    max-height: 33vh;
+    max-height: 100%;
+    will-change: z-index; /* Workaround for <rdar://125601926> - Content renders on top of the scroll bar */
```
`will-change: z-index` is a workaround for the same issue as #28050 where the scrollbar of the scroll container is obscured by the content within.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Main.css:
```diff
@@ -175,6 +175,7 @@ body.docked:is(.right, .left) #navigation-sidebar.collapsed > .resizer {
 #content {
     display: grid;
     height: 100%; /* This reduces paint areas when typing in the console. http://webkit.org/b/145324 */
+    flex: 1;
```
when did this break?  can we figure out the regression point and understand if this is something we caused or a change to something underneath us (e.g. WebCore)?  if the latter, id be concerned about more general webcompat issues

rcaliman-apple replied:
This regressed in https://commits.webkit.org/266784@main

[The changes in `Main.css`](https://github.com/WebKit/WebKit/commit/cfbd8f5ff5d3376688299cfe672e0bba44725372#diff-0ac5431454429c67756d9554bbc2bfe142df412a37b5503ea3eb2f1b6e21f76eL177) impacted the split-view Console before. We adjusted the maximum available space for the Console drawer in https://github.com/WebKit/WebKit/pull/27939/ to account for `min-height: 110px`  of  `#tab-browser`. But we didn't consider a multi-line Console prompt in that patch.

I tested the layout with the current patch and can't notice any negative impact on the work done in https://commits.webkit.org/266784@main.

rcaliman-apple replied:
@dcrousso Do you still have questions about this patch?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js:
```diff
@@ -154,7 +154,7 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
     {
         const minimumHeight = 64;
-        const maximumHeight = this.element.parentNode.offsetHeight - WI.TabBrowser.MinimumHeight - WI.QuickConsole.MinimumHeight;
+        const maximumHeight = this.element.parentNode.offsetHeight - WI.TabBrowser.MinimumHeight - WI.quickConsole.element.offsetHeight;
```
can you explain more the purpose of this specific part?

Style: i think this is technically a layering violation, so we should figure out a different way to access this
Style: this should be `let` as the (computed) value changes each time it's created (i.e. it's not truly constant)

rcaliman-apple replied:
> can you explain more the purpose of this specific part?

When resizing the Console drawer, we need to account for the min-height of the Tab content area (`WI.TabBrowser.MinimumHeight`) and for the height of the Console prompt. 

When there is multi-line input in the Console prompt, the _intent_ is to allow it to grow up to `33vh`. 
The drawer, prompt and Tab content area are all flex items so the fixed height of the drawer reduces from the available space to be distributed to the other items.

Without this change, the Console drawer can steal from the prompt's height. 
Here's a demo of the behavior (left: this patch, right: current Safari):

https://github.com/user-attachments/assets/a3874df1-a986-4836-b476-1200c797dd21


> Style: i think this is technically a layering violation, so we should figure out a different way to access this

Yeah, it's not ideal but they are semantically related. Do you have a suggestion how to separate?

**Later edit**: do you mean we shouldn't access and `this.element.parentNode` and `element.offsetHeight` directly and have getters instead?

dcrousso replied:
ok yeah i can see why that's useful/required

i was specifically referring to accessing the `WI.quickConsole` instance

it looks like we already calculate the `quickConsoleHeight` inside `WI.ConsoleDrawer.prototype._consoleResizerMouseDown`.  rather than fetch the `offsetHeight` on _every_ user action, could we save that value to a variable and use that instead?

rcaliman-apple replied:
> it looks like we already calculate the quickConsoleHeight inside WI.ConsoleDrawer.prototype._consoleResizerMouseDown. rather than fetch the offsetHeight on every user action, could we save that value to a variable and use that instead?

I tried this and I have some concerns about it.

```diff
diff --git a/Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js b/Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js
index 3940ba737761..67caa4002819 100644
--- a/Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js
+++ b/Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js
@@ -89,6 +89,13 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
         return this.element.offsetHeight;
     }
 
+    get quickConsoleHeight()
+    {
+        this._quickConsoleHeight ??= window.innerHeight - (this.element.totalOffsetTop + this.height);
+
+        return this._quickConsoleHeight;
+    }
+
     // Protected
 
     attached()
@@ -105,6 +112,8 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
         if (this._collapsed)
             return;
 
+        this._quickConsoleHeight = null;
+
         let height = this.height;
         this._restoreDrawerHeight();
 
@@ -124,8 +133,7 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
             return;
 
         let resizerElement = event.target;
-        let quickConsoleHeight = window.innerHeight - (this.element.totalOffsetTop + this.height);
-        let mouseOffset = quickConsoleHeight - (event.pageY - resizerElement.totalOffsetTop);
+        let mouseOffset = this.quickConsoleHeight - (event.pageY - resizerElement.totalOffsetTop);
 
         function dockedResizerDrag(event)
         {
@@ -140,6 +148,8 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
             if (event.button !== 0)
                 return;
 
+            this._quickConsoleHeight = null;
+
             WI.elementDragEnd(event);
         }
 
@@ -154,7 +164,7 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
     _updateDrawerHeight(height)
     {
         const minimumHeight = 64;
-        let maximumHeight = this.element.parentNode.offsetHeight - WI.TabBrowser.MinimumHeight - WI.quickConsole.element.offsetHeight;
+        let maximumHeight = this.element.parentNode.offsetHeight - WI.TabBrowser.MinimumHeight - this.quickConsoleHeight;
 
         let heightCSSValue = Number.constrain(height, minimumHeight, maximumHeight) + "px";
         if (this.element.style.height === heightCSSValue)
```

We need to compute the value of `this._quickConsoleHeight` independent of `WI.ConsoleDrawer._consoleResizerMouseDown()` because it is needed when restoring the drawer height when opening Web Inspector in `WI.ConsoleDrawer._restoreDrawerHeight()`.

We get an incorrect size if we lazy compute and cache the value on the first call:
`this._quickConsoleHeight ??= window.innerHeight - (this.element.totalOffsetTop + this.height);`

We _assume_ this is the height of the quick console, but during the initial layout stages, `this.element.totalOffsetTop + this.height` will be `0` so we end up with a wrong value: a quick console the full height of the viewport. 

We have to clear this value both in `WI.ConsoleDrawer.sizeDidChange()`, as the view structure is laid out, and after stopping dragging the console drawer resizer.

I think this makes it unnecessarily complex to reason about is prone to bugs in the future.

I agree that it's technically a layering violation, but I'd rather keep the direct access to `WI.quickConsole.element.offsetHeight` for clarity and for correctness.

rcaliman-apple replied:
@dcrousso do you agree that we can keep `WI.quickConsole.element.offsetHeight` for clarity and for correctness?

dcrousso replied:
my concern with this is that there's already some performance issues with resizing the Console (and Web Inspector in general) and i'd hate to make the situation worse by accessing values that force a relayout (if dirty)

is there truly no way of grabbing this value just once right when dragging starts (or maybe even shortly thereafter) so that we dont have to worry about perf?

what if instead of caching it in a `_quickConsoleHeight` property, we newly compute it each time in `_consoleResizerMouseDown` and `_restoreDrawerHeight` to then pass it as an argument to `_updateDrawerHeight`?

rcaliman-apple replied:
> what if instead of caching it in a _quickConsoleHeight property, we newly compute it each time in _consoleResizerMouseDown and _restoreDrawerHeight to then pass it as an argument to _updateDrawerHeight?

That's a reasonable approach 👍  
Added in the latest patch.

> my concern with this is that there's already some performance issues with resizing the Console (and Web Inspector in general) and i'd hate to make the situation worse by accessing values that force a relayout (if dirty)

Agreed. 

The issue that affects performance most here seems to be the resizing the Web Inspector window which will trigger `WI.ConsoleDrawer.sizeDidChange()` frequently and that always recalculates the available space for the drawer in `_restoreDrawerHeight()`. 

But that's a separate issue, ideally best resolved in CSS.

burg reviewed (approved):
r=me

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js:
```diff
@@ -148,13 +148,13 @@ WI.ConsoleDrawer = class ConsoleDrawer extends WI.ContentBrowser
+    _updateDrawerHeight(height, quickConsoleHeight)
     {
         const minimumHeight = 64;
-        const maximumHeight = this.element.parentNode.offsetHeight - WI.TabBrowser.MinimumHeight - WI.QuickConsole.MinimumHeight;
```
i dont think `WI.QuickConsole.MinimumHeight` is used anywhere else, so can we remove it?

or maybe we need to `Math.max(quickConsoleHeight, WI.QuickConsole.MinimumHeight)` instead (or more ideally before it's even passed to `_updateDrawerHeight` so that we don't have to recalculate it each time)?  if not, do we also need to update any of the uses of `--console-prompt-min-height`?

rcaliman-apple replied:
>i dont think WI.QuickConsole.MinimumHeight is used anywhere else, so can we remove it?

Yes, that member is obsolete now. Removed and updated the stylesheet.

>or maybe we need to Math.max(quickConsoleHeight, WI.QuickConsole.MinimumHeight) instead

No need to `Math.max()` that. The style ensures the minimum height ...

```css
.quick-console {
    min-height: var(--console-prompt-min-height);
}
```

... and that's going to be reflected in `WI.quickConsole.element.offsetHeight`.

>if not, do we also need to update any of the uses of --console-prompt-min-height?

There's a bunch of uses for `var(--console-prompt-min-height)` that are either obsolete or not really related to the Console prompt size. I filed a followup: https://bugs.webkit.org/show_bug.cgi?id=279436

rcaliman-apple committed:
Web Inspector: Multiline REPL editor is broken
https://bugs.webkit.org/show_bug.cgi?id=276530
rdar://131756916

Reviewed by BJ Burg and Devin Rousso.

The content area of the console prompt is controlled by CodeMirror
which sets the dimensions of a multi-line input as you type.
It lives in a second level flex container and is intended to be capped at 33vh.

At the first level flex container, the space available to be distributed to flex items
is influenced by the size of the Console Drawer and the min-width of #tab-browser.

Due to alignment within the flex layout and insufficient space,
the multi-line console prompt nested within the second level flex container
was allowed to grow within the second-level container without causing overflow.
This masks some keyboard input and focus.

This patch ensures the maximum height of the Console Drawer takes into account the
space used by the console prompt. It also moves the 33vh max-height limit
to the first level flex container to enforce the limit and ensure a scroll container
as soon as its contents overflow.

* Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.js:
(WI.ConsoleDrawer.prototype._updateDrawerHeight):
* Source/WebInspectorUI/UserInterface/Views/Main.css:
(#content):
* Source/WebInspectorUI/UserInterface/Views/QuickConsole.css:
(.quick-console):
(.quick-console > .console-prompt):

Canonical link: https://commits.webkit.org/283412@main
